src/axom/sidre/core/ConduitMemory.hpp

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2017-2025, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/*!
+ ******************************************************************************
+ *
+ * \file ConduitMemory.hpp
+ *
+ * \brief   Call-backs for using Axom memory management in Conduit.
+ *
+ ******************************************************************************
+ */
+
+#ifndef SIDRE_CONDUITMEMORY_HPP_
+#define SIDRE_CONDUITMEMORY_HPP_
+
+// Standard C++ headers
+#include <string>
+#include <set>
+#include <memory>
+
+#include "axom/config.hpp"
+#include "axom/core/memory_management.hpp"
+#include "axom/core/utilities/Utilities.hpp"
+#include "conduit_node.hpp"
+#include "conduit_utils.hpp"
+
+namespace axom
+{
+namespace sidre
+{
+
+/*!
+  @brief Object to do Conduit memory operations through Axom.
+
+  This class has no public constructor.  Use instanceForAxomId(int
+  axomAllocId) to access the instance for a specific Axom allocator
+  id.  The construction registers the appropriate callbacks with
+  Conduit, including the required memset and memcopy callbacks.
+
+  Allocator ids have a 1-to-1 relationship with allocators.
+
+  Axom's allocator is an extension of the Umpire allocator
+  when Umpire is used.  Conduit's allocator is opaque, but when used
+  by this class, it is associated with an Axom allocator (which is
+  an Umpire allocator).
+
+  Examples for setting Conduit allocator ids when you have Axom
+  allocator ids:
+
+  @code{.cpp}
+    void foo(conduit::Node& n, int axomAllocId) {
+      n.set_allocator(axomAllocIdToConduit(axomAllocId));
+    }
+
+    void bar(conduit::Node& n, int axomAllocId) {
+      const auto& instance = getInstance(axomAllocId);
+      assert(instance.axomId() == axomAllocId);
+      n.set_allocator(instance.conduitId());
+    }
+  @endcode
+*/
+struct ConduitMemory
+{
+  //!@brief Return the Axom allocator id.
+  int axomId() const { return m_axomId; }
+
+  //!@brief Return the Conduit allocator id coresponding to axomId().
+  conduit::index_t conduitId() const { return m_conduitId; }
+
+  /*!
+    @brief Convert an Axom allocator id to Conduit, registering
+    a new Conduit allocator if needed.
+  */
+  static conduit::index_t axomAllocIdToConduit(int axomAllocId)
+  {
+    return instanceForAxomId(axomAllocId).conduitId();
+  }
+
+  /*!
+    @brief Convert a Conduit allocator id to Axom.
+
+    The allocator must have been registered by a prior
+    instanceForAxomId() call.
+  */
+  static int conduitAllocIdToAxom(conduit::index_t conduitAllocId)
+  {
+    return instanceForConduitId(conduitAllocId).axomId();
+  }
+
+  /*!
+    @brief Return the instance for the given Axom allocator id.
+
+    This method IS NOT thread safe for new values of @c axomAllocId.
```

Comment:
Do we need it to be? It would be easy enough tot do that if necessary.

Reply:
It doesn't need to be.  Hard to come up with a scenario where you'd have multiple threads executing this part of the code.